### PR TITLE
DFAST-74: improve code coverage of the io

### PR DESCRIPTION
**Pull request by ralph-peelen-deltares**

### CODE REVIEW

File: tests/files/read_xyc_test.xyc [hunk 1]
```diff
@@ -0,0 +1,4 @@
+1 2 3
```
Comment (MRVermeulenDeltares) [inline tests/files/read_xyc_test.xyc:1]: Maybe we can look into using io for tests using small files.
I looked it up and it seems you can "Mock"  files.
https://docs.python.org/3/library/io.html

Comment (MRVermeulenDeltares) [reply tests/files/read_xyc_test.xyc:1]: example: 

```
    def test_read_waqua_xyz_Marlon_io_test(self):
        """
        Read WAQUA xyz file columns 1 and 2.
        """
        data_string = '''\
                        "A", "B", "C"
                        1, 2, 3
                        4, 5, 6
                        7, 8, 9
                        10, 11, 12
                        '''
        string_io = io.StringIO(data_string)
        string_io.name = "read_waqua_xyz_test.xyc"
        col = (1,2)
        data = dfastmi.io.read_waqua_xyz(string_io, col)
        datar = numpy.array([[ 2., 3.], [ 5., 6.], [ 8., 9.], [11., 12.]])
        print("data reference: ", datar)
        print("data read     : ", data)
        assert numpy.shape(data) == (4,2)
        assert (data == datar).all() == True
```

File: dfastmi/io.py [hunk 2]
```diff
@@ -1381,9 +808,25 @@ def read_xyc(
             colnames = ["Val", "X", "Y"]
         else:
             colnames = ["X", "Y"]
-        P = pandas.read_csv(
-            filename, names=colnames, skipinitialspace=True, delim_whitespace=True
-        )
+        if hasHeader:
```
Comment (MRVermeulenDeltares) [inline dfastmi/io.py:811]: Are these functional changes to this method?

File: dfastmi/io.py [hunk 3]
```diff
@@ -28,6 +28,9 @@
 """
 
 from typing import Tuple, Any, List, Union, Dict, Optional, TypedDict, TextIO
+import fiona
```
Comment (MRVermeulenDeltares) [inline dfastmi/io.py:31]: Why is this needed?

File: dfastmi/io.py [hunk 4]
```diff
@@ -36,37 +39,9 @@
 import pathlib
 import pandas
 import geopandas
-import zlib
 import shapely
-from packaging import version
 
 DataField = List[List[Union[float, List[float]]]]
-# RiversObject = Dict[str, Union[int, List[str], List[List[str]], DataField]]
```
[the remark this answers is not in the document]
Comment (MRVermeulenDeltares) [reply dfastmi/io.py:44]: discussed, will leave it in this PR, river files will be reviewed in refactor issue.

File: dfastmi/io.py [hunk 5]
```diff
@@ -196,354 +171,6 @@ def get_text(key: str) -> List[str]:
...
-                    )
-                fval = vals[0]
-            values_per_branch.append(fval)
-
-        all_values.append(values_per_branch)
-
-    return all_values
-
-
 def collect_int_values1(
```
Comment (MRVermeulenDeltares) [inline dfastmi/io.py:174]: This should also be in riverobject I think, test related to this can then also be moved.

File: dfastmi/io.py (hunk 2 above)
Comment (MRVermeulenDeltares) [reply dfastmi/io.py:811]: Discussed, this is needed for read_xyc to function with a correct file instead of a user changed file.

File: tests/test_io.py [hunk 6]
```diff
@@ -30,8 +30,21 @@ def test_load_program_texts_01(self):
...
+    def test_load_program_texts_02(self):
+        """
+        Testing load_program_texts.
+        """
+        print("current work directory: ", os.getcwd())
+        with pytest.raises(Exception) as cm:
+            dfastmi.io.load_program_texts("tests/files/messages.duplicate_keys.ini")
+        assert str(cm.value) == 'Duplicate entry for "checksum" in "tests/files/messages.duplicate_keys.ini".'
+
+@pytest.fixture
```
Comment (MRVermeulenDeltares) [inline tests/test_io.py:42]: you can move this up

File: tests/test_io.py [hunk 7]
```diff
@@ -84,6 +97,7 @@ def test_get_filename_01(self):
         """
         Testing get_filename wrapper for get_text.
         """
+        dfastmi.io.load_program_texts("dfastmi/messages.UK.ini")
```
Comment (MRVermeulenDeltares) [inline tests/test_io.py:100]: you can add the setup_data in this test and remove the"dfastmi.io.load_program_texts("dfastmi/messages.UK.ini")"

File: dfastmi/RiversObject.py [hunk 8]
```diff
@@ -0,0 +1,569 @@
...
+    prop_q: List[List[Tuple[float, ...]]]
+    prop_c: List[List[Tuple[float, ...]]]
+    proprate_high: List[List[float]] # only for version = 1
+    proprate_low: List[List[float]] # only for version = 1
+    qbankfull: List[List[float]] # only for version = 1
+    qmin: List[List[float]] # only for version = 1
+    qlevels: List[List[Tuple[float, float, float, float]]] # only for version = 1
+    dq: List[List[Tuple[float, float]]] # only for version = 1
+
+def read_rivers(filename: str = "rivers.ini") -> RiversObject:
```
Comment (MRVermeulenDeltares) [inline dfastmi/RiversObject.py:32]: From cmd.py a call to this method is made, but it is incorrect, this needs to be fixed, see:
"rivers = dfastmi.io.read_rivers(abs_rivers_file)"

File: dfastmi/io.py (hunk 5 above)
Comment (MRVermeulenDeltares) [reply dfastmi/io.py:174]: discussed will be tackeled in another issue

File: tests/test_io.py [hunk 9]
```diff
@@ -482,6 +379,456 @@ def test_write_simona_box_03(self):
         self.maxDiff = None
         assert all_lines == all_lines_ref
 
+class Test_collect_int_values1():
+    def test_collect_int_values1_01(self):
+        """
+
```
Comment (MRVermeulenDeltares) [inline tests/test_io.py:385]: Please either add something in the description or leave it out for now

Comment (MRVermeulenDeltares) [reply tests/test_io.py:385]: same for the other tests below

File: tests/test_io.py [hunk 10]
```diff
@@ -482,6 +379,456 @@ def test_write_simona_box_03(self):
...
...55.347 473982.089 16.053, 243257.378 473984.914 16.053, 243258.357 473987.727 16.053, 243258.67 473990.115 16.053, 243258.419 473992.703 16.053, 243257.297 473995.496 16.053, 243253.392 474001.476 16.053, 243246.321 474011.882 16.053, 243245.191 474013.607 16.053, 243244.65 474014.155 16.053, 243243.841 474014.667 16.053, 243242.965 474015.107 16.053, 243242.064 474015.309 16.053, 243240.948 474015.362 16.053, 243228.177 474012.682 16.053, 243208.861 474008.521 16.053, 243203.992 474005.365 16.053, 243197.289 474015.028 16.053), (243249.909 473975.707 16.053, 243244.279 473971.939 16.053, 243246.539 473968.78 16.053, 243252.068 473972.481 16.053, 243251.43 473973.434 16.053, 243249.909 473975.707 16.053), (246028.494 474654.753 16.053, 246028.408 474652.848 16.053, 246036.929 474652.797 16.053, 246036.912 474654.747 16.053, 246028.494 474654.753 16.053), (246036.997 474624.863 16.053, 246028.615 474624.835 16.053, 246028.649 474622.922 16.053, 246037.058 474622.874 16.053, 246036.997 474624.863 16.053), (251334.175 474118.685 16.308, 251333.896 474117.375 16.308, 251348.432 474115.058 16.308, 251348.643 474116.379 16.308, 251334.27 474118.67 16.308, 251334.175 474118.685 16.308)))'
```
Comment (MRVermeulenDeltares) [inline tests/test_io.py:830]: are all these values needed for this test?

**Review by MRVermeulenDeltares — CHANGES_REQUESTED** [review]
There is a failing test, please have a look at that test

File: tests/files/read_xyc_test.xyc (hunk 1 above)
Comment (MRVermeulenDeltares) [reply tests/files/read_xyc_test.xyc:1]: will be tackled in https://issuetracker.deltares.nl/browse/DFAST-86

File: tests/test_io.py (hunk 6 above)
Comment (ralph-peelen-deltares) [reply tests/test_io.py:42]: will do in https://issuetracker.deltares.nl/browse/DFAST-86

File: dfastmi/RiversObject.py (hunk 8 above)
Comment (ralph-peelen-deltares) [reply dfastmi/RiversObject.py:32]: done, good find thanks!

File: tests/test_io.py (hunk 9 above)
Comment (ralph-peelen-deltares) [reply tests/test_io.py:385]: left it here as place holder for https://issuetracker.deltares.nl/browse/DFAST-86

**Comment by ralph-peelen-deltares** [comment]
Resolved failing test

File: tests/test_io.py (hunk 7 above)
Comment (ralph-peelen-deltares) [reply tests/test_io.py:100]: will resolve in https://issuetracker.deltares.nl/browse/DFAST-86
